cmd/encryption.go

```diff
@@ -0,0 +1,101 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rancher/rke/cluster"
+	"github.com/rancher/rke/hosts"
+	"github.com/rancher/rke/pki"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+func EncryptionCommand() cli.Command {
+	encryptFlags := []cli.Flag{
+		cli.StringFlag{
+			Name:   "config",
+			Usage:  "Specify an alternate cluster YAML file",
+			Value:  pki.ClusterConfig,
+			EnvVar: "RKE_CONFIG",
+		},
+	}
+	encryptFlags = append(encryptFlags, commonFlags...)
+	return cli.Command{
+		Name:  "encrypt",
+		Usage: "Manage cluster encryption provider keys",
+		Subcommands: cli.Commands{
+			cli.Command{
+				Name:   "rotate-key",
+				Usage:  "Rotate cluster encryption provider key",
+				Action: rotateEncryptionKeyFromCli,
+				Flags:  encryptFlags,
+			},
+		},
+	}
+}
+
+func rotateEncryptionKeyFromCli(ctx *cli.Context) error {
+	logrus.Infof("Running RKE version: %v", ctx.App.Version)
+	clusterFile, filePath, err := resolveClusterFile(ctx)
+	if err != nil {
+		return fmt.Errorf("Failed to resolve cluster file: %v", err)
+	}
+
+	rkeConfig, err := cluster.ParseConfig(clusterFile)
+	if err != nil {
+		return fmt.Errorf("Failed to parse cluster file: %v", err)
+	}
+
+	rkeConfig, err = setOptionsFromCLI(ctx, rkeConfig)
+	if err != nil {
+		return err
+	}
+
+	// setting up the flags
+	flags := cluster.GetExternalFlags(false, false, false, "", filePath)
+
+	return RotateEncryptionKey(context.Background(), rkeConfig, hosts.DialersOptions{}, flags)
+}
+
+func RotateEncryptionKey(ctx context.Context, rkeConfig *v3.RancherKubernetesEngineConfig,
+	dialersOptions hosts.DialersOptions, flags cluster.ExternalFlags) error {
+
+	stateFilePath := cluster.GetStateFilePath(flags.ClusterFilePath, flags.ConfigDir)
+	rkeFullState, _ := cluster.ReadStateFile(ctx, stateFilePath)
+	// We generate the first encryption config in ClusterInit, to store it ASAP. It's written
+	// to the DesiredState
+	stateEncryptionConfig := rkeFullState.DesiredState.EncryptionConfig
+
+	// if CurrentState has EncryptionConfig, it means this is NOT the first time we enable encryption, we should use the _latest_ applied value from the current cluster
+	if rkeFullState.CurrentState.EncryptionConfig != "" {
+		stateEncryptionConfig = rkeFullState.CurrentState.EncryptionConfig
+	}
+
+	kubeCluster, err := cluster.InitClusterObject(ctx, rkeConfig, flags, stateEncryptionConfig)
```

Comment:
@moelsayed why stateEncryption config is not a part of rke config?

Reply:
We try to keep everything in the state, not the config. This way we can persist it faster and avoid syncing issues across kontainer-engine when working with rancher. Additional, in rancher case, the configuration will be stored in the cluster store in a secret, which should be more secure than the cluster CRD.